presto-parser/src/main/java/com/facebook/presto/sql/tree/CopyQueryVisitor.java

```diff
@@ -0,0 +1,325 @@
+    {
+        if (node.getLocation().isPresent()) {
+            return new Query(
+                    node.getLocation().get(),
+                    node.getWith().map(this::processSameType),
+                    processSameType(node.getQueryBody()),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+        else {
+            return new Query(
+                    node.getWith().map(this::processSameType),
+                    processSameType(node.getQueryBody()),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, C context)
+    {
+        if (node.getLocation().isPresent()) {
+            return new QuerySpecification(
+                    node.getLocation().get(),
+                    node.getSelect(),
+                    node.getFrom().map(this::processSameType),
+                    node.getWhere(),
+                    node.getGroupBy(),
+                    node.getHaving(),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+        else {
+            return new QuerySpecification(
+                    node.getSelect(),
+                    node.getFrom().map(this::processSameType),
+                    node.getWhere(),
+                    node.getGroupBy(),
+                    node.getHaving(),
+                    node.getOrderBy(),
+                    node.getOffset(),
+                    node.getLimit());
+        }
+    }
+
+    @Override
+    protected Node visitUnion(Union node, C context)
+    {
+        List<Relation> newRelations = processNodeList(node.getRelations());
+        if (node.getLocation().isPresent()) {
+            return new Union(
+                    node.getLocation().get(),
+                    newRelations,
+                    node.isDistinct());
+        }
+        else {
+            return new Union(
+                    newRelations,
+                    node.isDistinct());
+        }
+    }
+
+    @Override
+    protected Node visitIntersect(Intersect node, C context)
+    {
+        List<Relation> newRelations = processNodeList(node.getRelations());
+
+        if (node.getLocation().isPresent()) {
+            return new Intersect(
+                    node.getLocation().get(),
+                    newRelations,
+                    node.isDistinct());
+        }
+        else {
+            return new Intersect(
+                    newRelations,
+                    node.isDistinct());
+        }
+    }
+
+    @Override
+    protected Node visitExcept(Except node, C context)
+    {
+        Relation left = processSameType(node.getLeft());
+        Relation right = processSameType(node.getRight());
+
+        if (node.getLocation().isPresent()) {
```

Comment:
Its sad we need to do this if else. We can add a todo to consider changing constructor definitions for all classes extending Node. We can make constructor with Optional<SourceLocation> public and removing the constructor with SourceLocation. It will be a big refactor so adding a todo in this class, and doing later is fine. Cc @highker for his thoughts

Reply:
+1, I don't personally see how an `Optional` constructor would cause much of a problem, outside of an `Optional used as type parameter` warning. Happy to spend 2 hours making a python script to fix this instead of spending 1 hour doing it manually ;)